src/trace.rs

```diff
@@ -0,0 +1,304 @@
+    }
+
+    let subscriber = Registry::default().with(layers);
+
+    tracing::subscriber::set_global_default(subscriber)?;
+
+    // Install a logger that converts logs into tracing events
+    LogTracer::init()?;
+
+    Ok((
+        TraceGuards {
+            uses_otel_tracer: config.open_telemetry_config.is_some(),
+            _chrome_guard: chrome_guard,
+        },
+        stdout_filter_handle,
+    ))
+}
+
+#[allow(missing_debug_implementations)]
+pub struct TraceGuards {
+    uses_otel_tracer: bool,
+    _chrome_guard: Option<tracing_chrome::FlushGuard>,
+}
+
+impl Drop for TraceGuards {
+    fn drop(&mut self) {
+        if self.uses_otel_tracer {
+            // Flush buffered traces in the OpenTelemetry pipeline.
+            opentelemetry::global::shutdown_tracer_provider();
+        }
+    }
+}
+
+pub fn traceconfig_handler(trace_reload_handle: TraceReloadHandle) -> impl Handler {
+    (
+        State(Arc::new(trace_reload_handle)),
+        Router::new()
+            .get("/traceconfig", api(get_traceconfig))
+            .put("/traceconfig", api(put_traceconfig)),
+    )
+}
+
+async fn get_traceconfig(
+    conn: &mut trillium::Conn,
+    State(trace_reload_handle): State<Arc<TraceReloadHandle>>,
+) -> Result<Json<TraceconfigBody>, Status> {
+    Ok(Json(TraceconfigBody {
+        filter: trace_reload_handle
+            .with_current(|trace_filter| trace_filter.to_string())
+            .map_err(|err| {
+                conn.set_body(format!("failed to get current filter: {err}"));
+                Status::InternalServerError
+            })?,
+    }))
+}
+
+/// Allows modifying the runtime tracing filter. Accepts a request with a body corresponding to
+/// [`TraceconfigBody`]. If the `filter` field is empty, the filter will fallback to `error`.
+/// See [`EnvFilter::try_new`] for details.
+async fn put_traceconfig(
+    conn: &mut trillium::Conn,
+    (State(trace_reload_handle), Json(req)): (State<Arc<TraceReloadHandle>>, Json<TraceconfigBody>),
+) -> Result<Json<TraceconfigBody>, Status> {
+    let new_filter = EnvFilter::try_new(req.filter).map_err(|err| {
+        conn.set_body(format!("invalid filter: {err}"));
+        Status::BadRequest
+    })?;
+    trace_reload_handle.reload(new_filter).map_err(|err| {
+        conn.set_body(format!("failed to update filter: {err}"));
+        Status::InternalServerError
+    })?;
+    Ok(Json(TraceconfigBody {
+        filter: trace_reload_handle
+            .with_current(|trace_filter| trace_filter.to_string())
+            .map_err(|err| {
+                conn.set_body(format!("failed to get current filter: {err}"));
+                Status::InternalServerError
+            })?,
+    }))
+}
+
+/// The response and request body used by /traceconfig for reporting and updating its configuration.
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+struct TraceconfigBody {
+    /// The directive that filters spans and events. This field follows the [`EnvFilter`][1] syntax.
```

Comment:
My thought was for future expandability of this endpoint of other parameters we'd want to tweak at runtime, but YAGNI might prevail here (and even then, it might be better to have subpaths of `/traceconfig` rather than JSON wrangling).

I'm OK with changing to a simple string body (I would have to backport this to Janus as well for consistency).